### PR TITLE
Cleanup timeouts in pthread condvar

### PR DESCRIPTION
**Pull request by joboet**

### CODE REVIEW

File: library/std/src/sys/unix/locks/pthread_condvar.rs [hunk 1]
```diff
@@ -110,25 +104,14 @@ impl Condvar {
...
-            sec.map(|s| libc::timespec { tv_sec: s, tv_nsec: nsec as _ }).unwrap_or(TIMESPEC_MAX);
+        use crate::sys::time::Timespec;
 
+        let timeout = Timespec::now(libc::CLOCK_MONOTONIC)
+            .checked_add_duration(&dur)
+            .and_then(|t| t.to_timespec())
+            .unwrap_or(TIMESPEC_MAX);
         let r = libc::pthread_cond_timedwait(self.inner.get(), pthread_mutex::raw(mutex), &timeout);
-        assert!(r == libc::ETIMEDOUT || r == 0);
+        debug_assert!(r == libc::ETIMEDOUT || r == 0);
```
Comment (Mark-Simulacrum) [inline library/std/src/sys/unix/locks/pthread_condvar.rs:114]: Why the change from assert! here?

File: library/std/src/sys/unix/locks/pthread_condvar.rs [hunk 2]
```diff
@@ -143,56 +126,38 @@ impl Condvar {
...
             .unwrap_or(TIMESPEC_MAX);
-
-        // And wait!
         let r = libc::pthread_cond_timedwait(self.inner.get(), pthread_mutex::raw(mutex), &timeout);
         debug_assert!(r == libc::ETIMEDOUT || r == 0);
 
         // ETIMEDOUT is not a totally reliable method of determining timeout due
         // to clock shifts, so do the check ourselves
-        stable_now.elapsed() < dur
+        r == 0 || now.elapsed() < dur
```
Comment (Mark-Simulacrum) [inline library/std/src/sys/unix/locks/pthread_condvar.rs:160]: Why this change?

File: library/std/src/sys/unix/locks/pthread_condvar.rs (hunk 1 above)
Comment (joboet) [reply library/std/src/sys/unix/locks/pthread_condvar.rs:114]: Because all other errors are only emitted by some implementations, but the incorrect usage they are indicating may also cause UB. Hence, this assertion is only really useful as a simple smoke test in debug builds. This also matches the assert in `wait`.

File: library/std/src/sys/unix/locks/pthread_condvar.rs (hunk 2 above)
Comment (joboet) [reply library/std/src/sys/unix/locks/pthread_condvar.rs:160]: This saves the thread from having to check the time when it has received a wakeup. This is a very slight change in behaviour, however, so I can definitely move this change to a follow-up PR.

Comment (Mark-Simulacrum) [reply library/std/src/sys/unix/locks/pthread_condvar.rs:160]: I don't have bandwidth right now to evaluate this change with confidence so I would appreciate splitting it out. I'm not sure we can rely on wakeups being non-spurious, and your suggestion of change in behavior definitely seems to support that.

File: library/std/src/sys/unix/locks/pthread_condvar.rs (hunk 1 above)
Comment (Mark-Simulacrum) [reply library/std/src/sys/unix/locks/pthread_condvar.rs:114]: I personally feel like we should keep this as an assert, but I'm not sure what the current consensus there is - I know there was some discussion recently about this....

https://github.com/rust-lang/rust/pull/99624 seems like it generally moves us in the direction of preferring assert, and I would want that myself, so I think we should stick with asserts.

Comment (joboet) [reply library/std/src/sys/unix/locks/pthread_condvar.rs:114]: Should I update the other assertions too?

Comment (joboet) [reply library/std/src/sys/unix/locks/pthread_condvar.rs]: I've now kept all assertions as they were.

**Comment by joboet** [comment]
@rustbot ready
@Mark-Simulacrum I've now kept all assertions and the behaviour of timeouts the same, so this PR now only contains the cleanup.
